### ISSUE TITLE
Unspreadable type but no error

### ISSUE THREAD

**Issue by liuxilu** [microsoft/TypeScript#56750]
### 🔎 Search Terms

array spread operator type check, unspreadable type, array spread

### 🕗 Version & Regression Information

5.4.0-dev.20231212.

### ⏯ Playground Link

https://www.typescriptlang.org/play?ts=5.4.0-dev.20231212#code/MYGwhgzhAEByCuBbaBvaBfAUAFwJ4AcBTOJGAXlWgDtSAuExAbQF0NMRDtox6AKBRBAA+AgJQtoFRlUIB3Br1HMA3JjAA6AE6EAJvGCFevbPQEsANNEKjJAPmjH1+eBAAWvdZ97FIDCKPUaQWghIStRS2wI6BZRTCA

### 💻 Code

```ts
class Num { }
type Nums = { nums: Num[] }
let a: (Nums|Num)[] = [new Num()];
a.reduce((t: Num[], e) => (t.push(...(e as Nums).nums || e), t), [])
```

### 🙁 Actual behavior

No error and runtime crash.

### 🙂 Expected behavior

Type check error. Because `Num` is unspreadable.

**Comment by MartinJohns** [comment]
You get an error, but you tell the compiler to ignore it using the type assertion `as`.

`e` can either be an array, or it can be a `Nums`. When it's an array, accessing `nums` results in `undefined`. After this you use the or-operator with `e`, and since `undefined` is falsy you end up with `e`, but `e` is actually a `Num` and can't be spread, resulting in your error.

The compiler does not complain further, because `nums` is typed `Num[]` and can never be falsy (according to the type system), so the right side type is irrelevant. If you disable `strictNullChecks` another error is reported.

**Comment by liuxilu** [comment]
I see, thank you.
